crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs

```diff
@@ -963,10 +981,67 @@ impl<'p> DeconstructedPat<'p> {
         DeconstructedPat::new(ctor, fields, pat.ty.clone())
     }
 
-    // // FIXME(iDawer): implement reporting of noncovered patterns
-    // pub(crate) fn to_pat(&self, _cx: &MatchCheckCtx<'_, 'p>) -> Pat {
-    //     Pat { ty: self.ty.clone(), kind: PatKind::Wild.into() }
-    // }
+    pub(crate) fn to_pat(&self, cx: &MatchCheckCtx<'_, 'p>) -> Pat {
+        let mut subpatterns = self.iter_fields().map(|p| p.to_pat(cx));
+        let pat = match &self.ctor {
+            Single | Variant(_) => match self.ty.kind(Interner) {
+                TyKind::Tuple(..) => PatKind::Leaf {
+                    subpatterns: subpatterns
+                        .zip(0u32..)
+                        .map(|(p, i)| FieldPat {
+                            field: LocalFieldId::from_raw(i.into()),
+                            pattern: p,
+                        })
+                        .collect(),
+                },
+                TyKind::Adt(adt, _) if is_box(adt.0, cx.db) => {
+                    // Without `box_patterns`, the only legal pattern of type `Box` is `_` (outside
+                    // of `std`). So this branch is only reachable when the feature is enabled and
+                    // the pattern is a box pattern.
+                    PatKind::Deref { subpattern: subpatterns.next().unwrap() }
+                }
+                TyKind::Adt(adt, substs) => {
+                    let variant = self.ctor.variant_id_for_adt(adt.0);
+                    let subpatterns = Fields::list_variant_nonhidden_fields(cx, self.ty(), variant)
+                        .zip(subpatterns)
+                        .map(|((field, _ty), pattern)| FieldPat { field, pattern })
+                        .collect();
+
+                    if let VariantId::EnumVariantId(enum_variant) = variant {
+                        PatKind::Variant { substs: substs.clone(), enum_variant, subpatterns }
+                    } else {
+                        PatKind::Leaf { subpatterns }
+                    }
+                }
+                // Note: given the expansion of `&str` patterns done in `expand_pattern`, we should
+                // be careful to reconstruct the correct constant pattern here. However a string
+                // literal pattern will never be reported as a non-exhaustiveness witness, so we
+                // ignore this issue.
+                TyKind::Ref(..) => PatKind::Deref { subpattern: subpatterns.next().unwrap() },
+                _ => {
+                    never!("unexpected ctor for type {:?} {:?}", self.ctor, self.ty);
+                    PatKind::Wild
+                }
+            },
+            &Slice(Slice { _unimplemented: _void }) => unimplemented!(),
+            &Str(_void) => unimplemented!(),
+            &FloatRange(_void) => unimplemented!(),
```

Comment:
maybe use `never!` for these unimplementeds as well?

Reply:
I understand these as 'to be implemented' not as bugs. As these variants are being implemented the `unimplemented!()`'s supposed to catch eyes.
They're (almost) all behind void enum, practically shouldn't ever hit them at runtime.

Reply:
> practically shouldn't ever hit them at runtime

In other words, it's a bug if we ever hit them at runtime, right?

Can you instead `match void {}` to prove statically they won't be hit?

Reply:
Just to clarify, `never!` being in the code is not a bug, it's completely normal. `never!` *being hit at runtime* is a bug.

Reply:
> Can you instead `match void {}` to prove statically they won't be hit?

Yes, this is even better.

Reply:
Done, thank you!
Good for merge?